api/trace/testtrace/span_test.go

```diff
@@ -0,0 +1,558 @@
+		})
+
+		t.Run("cannot be changed after the span has been ended", func(t *testing.T) {
+			t.Parallel()
+
+			e := matchers.NewExpecter(t)
+
+			tracer := testtrace.NewTracer()
+			_, span := tracer.Start(context.Background(), "test")
+
+			subject, ok := span.(*testtrace.Span)
+			e.Expect(ok).ToBeTrue()
+
+			_, linkedSpan := tracer.Start(context.Background(), "linked")
+			expectedLink := trace.Link{
+				SpanContext: linkedSpan.SpanContext(),
+				Attributes: []core.KeyValue{
+					core.Key("key").String("value"),
+				},
+			}
+
+			subject.AddLink(expectedLink)
+			subject.End()
+
+			_, unlinkedSpan := tracer.Start(context.Background(), "not linked")
+			unexpectedLink := trace.Link{
+				SpanContext: unlinkedSpan.SpanContext(),
+			}
+
+			subject.AddLink(unexpectedLink)
+
+			links := subject.Links()
+
+			e.Expect(len(links)).ToEqual(1)
+			e.Expect(links[expectedLink.SpanContext]).ToEqual(expectedLink.Attributes)
+		})
+	})
+
+	t.Run("#Events", func(t *testing.T) {
+		t.Run("returns an empty slice by default", func(t *testing.T) {
+			t.Parallel()
+
+			e := matchers.NewExpecter(t)
+
+			tracer := testtrace.NewTracer()
+			_, span := tracer.Start(context.Background(), "test")
+
+			subject, ok := span.(*testtrace.Span)
+			e.Expect(ok).ToBeTrue()
+
+			e.Expect(len(subject.Events())).ToEqual(0)
+		})
+
+		t.Run("returns all of the added events", func(t *testing.T) {
+			t.Parallel()
+
+			e := matchers.NewExpecter(t)
+
+			tracer := testtrace.NewTracer()
+			_, span := tracer.Start(context.Background(), "test")
+
+			subject, ok := span.(*testtrace.Span)
+			e.Expect(ok).ToBeTrue()
+
+			event1Msg := "event1"
+			event1Attributes := []core.KeyValue{
+				core.Key("event1Attr1").String("foo"),
+				core.Key("event1Attr2").String("bar"),
+			}
+
+			event1Start := time.Now()
+			subject.AddEvent(context.Background(), event1Msg, event1Attributes...)
+			event1End := time.Now()
+
+			event2Timestamp := time.Now().AddDate(5, 0, 0)
+			event2Msg := "event1"
+			event2Attributes := []core.KeyValue{
+				core.Key("event2Attr").String("abc"),
+			}
+
+			subject.AddEventWithTimestamp(context.Background(), event2Timestamp, event2Msg, event2Attributes...)
+
+			events := subject.Events()
+
+			e.Expect(len(events)).ToEqual(2)
+
+			event1 := events[0]
+
+			e.Expect(event1.Timestamp).ToBeTemporally(matchers.AfterOrSameTime, event1Start)
+			e.Expect(event1.Timestamp).ToBeTemporally(matchers.BeforeOrSameTime, event1End)
+			e.Expect(event1.Message).ToEqual(event1Msg)
+
+			for _, attr := range event1Attributes {
+				e.Expect(event1.Attributes[attr.Key]).ToEqual(attr.Value)
+			}
+
+			event2 := events[1]
+
+			e.Expect(event2.Timestamp).ToEqual(event2Timestamp)
+			e.Expect(event2.Message).ToEqual(event2Msg)
+
+			for _, attr := range event2Attributes {
+				e.Expect(event2.Attributes[attr.Key]).ToEqual(attr.Value)
+			}
+		})
+
+		t.Run("cannot be changed after the span has been ended", func(t *testing.T) {
```

Comment:
I think this is a duplicate of the `Link` test and should probably be removed since there is a  `can't be changed after end` for both `Links` and `Events` already.

Reply:
Yup, looks like a dupe. I'll remove it 👍